### PR TITLE
feat: [CI-4810]: Adding OAuth option to Github connector

### PR DESCRIPTION
**Pull request by vardanbansal-harness**
##### Summary:

1. Added new `OAuth` authentication option, enabled only free users.

https://user-images.githubusercontent.com/48940704/175323648-b5746836-d215-4ce2-bc35-25c579ae245b.mov

2. Re-link with Git, in the event of access being revoked.

https://user-images.githubusercontent.com/48940704/175325456-a511966f-cd0d-494c-9e12-8b84ef6ff0cc.mov

##### Screenshots:

You can use the following comments to re-trigger PR Checks

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress: `trigger cypress`
- Fix Prettier - `fix prettier`

### CODE REVIEW

**Review by d1wilko — APPROVED** [review]
Some great code here - 2 suggestions - but approving regardless

File: src/modules/35-connectors/components/CreateConnector/GithubConnector/StepAuth/StepGithubAuthentication.tsx [hunk 1]
```diff
@@ -290,15 +428,59 @@ const StepGithubAuthentication: React.FC<StepProps<StepGithubAuthenticationProps
...
+                            const oAuthURL = await response.text()
+                            if (typeof oAuthURL === 'string') {
+                              window.open(oAuthURL, '_blank')
+                            }
+                          } catch (e) {
+                            setOAuthStatus(Status.FAILURE)
+                            clear()
+                            showError(getString('connectors.oAuth.failed'))
+                          }
+                        }}
```
[the remark this answers is not in the document]
Comment (vardanbansal-harness) [reply src/modules/35-connectors/components/CreateConnector/GithubConnector/StepAuth/StepGithubAuthentication.tsx]: Done.

**Comment by vardanbansal-harness** [comment]
> Some great code here - 2 suggestions - but approving regardless

Incorporated the suggestions, thanks @d1wilko !

**Review by upadhyay-an — DISMISSED** [review]

**Review by avinashmadhwani02 — APPROVED** [review]

File: src/modules/35-connectors/components/CreateConnector/CreateConnectorUtils.ts [hunk 2]
```diff
@@ -49,3 +50,19 @@ export const getCommonConnectorsValidationSchema = (
       otherwise: Yup.object().nullable()
     })
   })
+
+export const getBackendServerUrl = (): string => {
```
[the remark this answers is not in the document]
Comment (vardanbansal-harness) [reply src/modules/35-connectors/components/CreateConnector/CreateConnectorUtils.ts:54]: Specifically, it's only going to be used in Git connectors, and that too only for OAuth.

File: src/modules/35-connectors/components/CreateConnector/GithubConnector/StepAuth/StepGithubAuthentication.tsx [hunk 3]
```diff
@@ -166,13 +189,125 @@ const StepGithubAuthentication: React.FC<StepProps<StepGithubAuthenticationProps
...
+          window.open(oAuthURL, '_blank')
+        }
+      } catch (e) {
+        setOAuthStatus(Status.FAILURE)
+        clear()
+        showError(getString('connectors.oAuth.failed'))
+      }
+    }, [isExistingOAuthConnectionHealthy, accountId])
+
+    window.addEventListener('message', handleOAuthServerEvent)
```
[the remark this answers is not in the document]
Comment (vardanbansal-harness) [reply src/modules/35-connectors/components/CreateConnector/GithubConnector/StepAuth/StepGithubAuthentication.tsx:292]: Yes, but we do remove it right away once `oAuthSecretIntercepted.current` is set to true, as soon as first OAuth event is received from server.

File: src/modules/35-connectors/pages/connectors/utils/ConnectorUtils.ts [hunk 4]
```diff
@@ -197,6 +197,39 @@ const getGitAuthSpec = (formData: FormData) => {
...
+
+const getGitApiAccessSpec = (formData: FormData): Record<string, any> => {
+  const { authType = '' } = formData
+  switch (authType) {
+    case GitAuthTypes.OAUTH:
+      const oAuthAccessTokenRef =
+        formData.oAuthAccessTokenRef || get(formData, 'spec.authentication.spec.spec.tokenRef')
+      const oAuthRefreshTokenRef =
+        formData.oAuthRefreshTokenRef || get(formData, 'spec.authentication.spec.spec.refreshTokenRef')
+      return {
```
[the remark this answers is not in the document]
Comment (vardanbansal-harness) [reply src/modules/35-connectors/pages/connectors/utils/ConnectorUtils.ts:222]: Done.

**Review by avinashmadhwani02 — APPROVED** [review]

**Pull request merged**